### PR TITLE
Implementation of multi-stage process of switch rule installation.

### LINKED ISSUE #1711: Ensure new rules installed before removing old ones on reroute
Ensure new switch rules installed before removing old ones on flow reroute, in order to eliminate cases 
when a flow becomes broken due to unsuccessful rule setup.

### PR DESCRIPTION
**Pull request by sergii-iakovenko**
The changes introduce FlowCommandRegistry in FlowTopology. TransactionBolt utilizes it as an intermediate storage for commands during the multi-stage process. In accordance with #1712.

Resolves #1711

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/1748)

### CODE REVIEW

File: services/src/messaging/src/main/java/org/openkilda/messaging/command/flow/BatchFlowCommandsRequest.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+/**
+ * Defines the payload of a message representing a batch of grouped flow commands (install / remove). The batch must be
+ * either completely processed group by group or canceled as a whole.
+ */
+@Value
+public class BatchFlowCommandsRequest extends CommandData {
+    @JsonProperty("groups")
+    private List<FlowCommandGroup> groups;
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply services/src/messaging/src/main/java/org/openkilda/messaging/command/flow/BatchFlowCommandsRequest.java]: Fixed.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowCommandFactory.java [hunk 2]
```diff
@@ -29,46 +29,41 @@
...
-        OutputVlanType outputVlanType = getOutputVlanType(flow);
-
+    public List<InstallTransitFlow> createInstallTransitAndEgressRulesForFlow(Flow flow, List<FlowSegment> segments) {
         if (Objects.equals(flow.getSrcSwitch().getSwitchId(), flow.getDestSwitch().getSwitchId())) {
-            return Collections.singletonList(makeOneSwitchRule(flow, outputVlanType));
+            return Collections.emptyList();
         }
         if (segments.isEmpty()) {
-            throw new IllegalArgumentException("Neither one switch flow nor flow path.");
+            throw new IllegalArgumentException("Neither one switch flow nor flow segments provided");
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowCommandFactory.java]: Fixed.

File: services/wfm/src/test/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistryTest.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    @Test
+    public void shouldReturnFlowForExpiredBatch() throws InterruptedException {
+        FlowCommandRegistry registry = new FlowCommandRegistry();
+        registerBatchWith2Groups(registry);
+
+        registry.pollNextGroup(TEST_FLOW);
+
+        Thread.sleep(100);
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply services/wfm/src/test/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistryTest.java]: Fixed.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        }
+
+        batches.put(batchId, new Batch(flowId, batchTransactions));
+    }
+
+    /**
+     * Removes the command (identified by the flow and transaction IDs) from the current group.
+     * Cleans up groups and batches if they are empty.
+     */
+    public synchronized void removeCommand(String flowId, UUID transactionId) throws UnknownTransactionException {
```
Comment (surabujin) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java]: `synchnozized` in storm bolts? Why for?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        for (List<BaseFlow> groupCommands : batchCommands) {
+            flowGroups.add(new Group(batchId, new ArrayList<>(groupCommands)));
+
+            for (BaseFlow command : groupCommands) {
+                if (!command.getId().equals(flowId)) {
+                    throw new IllegalArgumentException(
+                            format("Command '%s' doesn't correspond to specified flow %s", command, flowId));
+                }
+
+                UUID transactionId = command.getTransactionId();
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java]: The check has been added into the registerBatch method.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java (hunk 4 above)
Comment (sergii-iakovenko) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java]: Good point. Removed and added notice to the class javadoc.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/bolts/TransactionBolt.java [hunk 6]
```diff
@@ -41,141 +54,163 @@
...
+                    logger.error("Unexpected component: {} in {}", componentId, tuple);
                     break;
             }
-        } catch (RuntimeException exception) {
-            logger.error("Set status {}: switch-id={}, {}={}, {}={}",
-                    FlowState.DOWN, switchId, Utils.FLOW_ID, flowId, Utils.TRANSACTION_ID, transactionId, exception);
 
-            values = new Values(flowId, FlowState.DOWN, new CommandContext());
-            outputCollector.emit(StreamType.STATUS.toString(), tuple, values);
+            checkForExpiredCommands(tuple, message.getCorrelationId());
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flow/bolts/TransactionBolt.java]: Fixed. Moved to doTick.

**Pull request merged**